spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/SecretManagerPropertySource.java

```diff
@@ -0,0 +1,112 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.secretmanager;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.cloud.secretmanager.v1beta1.AccessSecretVersionResponse;
+import com.google.cloud.secretmanager.v1beta1.ProjectName;
+import com.google.cloud.secretmanager.v1beta1.Secret;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient.ListSecretsPagedResponse;
+import com.google.cloud.secretmanager.v1beta1.SecretVersionName;
+import com.google.protobuf.ByteString;
+
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.core.env.EnumerablePropertySource;
+
+/**
+ * Retrieves secrets from GCP Secret Manager under the current GCP project.
+ *
+ * @author Daniel Zou
+ * @since 1.3
+ */
+public class SecretManagerPropertySource extends EnumerablePropertySource<SecretManagerServiceClient> {
+
+	private static final String LATEST_VERSION_STRING = "latest";
+
+	private final Map<String, Object> properties;
+
+	private final String[] propertyNames;
+
+	public SecretManagerPropertySource(
+			String propertySourceName,
+			SecretManagerServiceClient client,
+			GcpProjectIdProvider projectIdProvider,
+			String secretsNamespace) {
+
+		super(propertySourceName, client);
+
+		Map<String, Object> propertiesMap = initializePropertiesMap(
+				client, projectIdProvider.getProjectId(), secretsNamespace);
+
+		this.properties = propertiesMap;
+		this.propertyNames = propertiesMap.keySet().toArray(new String[propertiesMap.size()]);
+	}
+
+	@Override
+	public String[] getPropertyNames() {
+		return propertyNames;
+	}
+
+	@Override
+	public Object getProperty(String name) {
+		return properties.get(name);
+	}
+
+	private static Map<String, Object> initializePropertiesMap(
+			SecretManagerServiceClient client, String projectId, String secretsNamespace) {
+
+		ListSecretsPagedResponse response = client.listSecrets(ProjectName.of(projectId));
+
+		HashMap<String, Object> secretsMap = new HashMap<>();
+		for (Secret secret : response.iterateAll()) {
+			String secretId = extractSecretId(secret);
+			ByteString secretPayload = getSecretPayload(client, projectId, secretId);
+			secretsMap.put(secretsNamespace + secretId, secretPayload);
+		}
+
+		return secretsMap;
+	}
+
+	private static ByteString getSecretPayload(
+			SecretManagerServiceClient client, String projectId, String secretId) {
+
+		SecretVersionName secretVersionName = SecretVersionName.newBuilder()
+				.setProject(projectId)
+				.setSecret(secretId)
+				.setSecretVersion(LATEST_VERSION_STRING)
```

Comment:
I think people will want to access specific secret versions. Defaulting to "latest" is not a production best practice.

Reply:
Ah good point, we'll definitely need an answer for this. I think we'll have to discuss a bit more in-depth with our team what is the best way to let users specify which secret versions they want to load into the environment; I guess we wouldn't want to load all versions of all secrets by default.

I added this point to #2176 to track; we will definitely implement this before releasing to users. I just want this PR to setup the initial scaffolding.